core/src/main/scala/org/http4s/HeaderV2.scala

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2013 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.header.v2
+
+import cats.{Id, Monoid, Order, Show}
+import cats.data.NonEmptyList
+import cats.syntax.all._
+import org.typelevel.ci.CIString
+import scala.collection.mutable.{ListBuffer, Set => MutSet}
+
+/** Typeclass representing an HTTP header, which all the http4s
+  * default headers satisfy.
+  * You can add custom headers by providing an implicit instance of
+  * `Header[YourCustomHeader]`
+  */
+trait Header[A, T <: Header.Type] {
+
+  /** Name of the header. Not case sensitive.
+    */
+  def name: CIString
+
+  /** Value of the header, which is represented as a String.
+    * Will be a comma separated String for headers with multiple values.
+    */
+  def value(a: A): String
+
+  /** Parses the header from its String representation.
+    * Could be a comma separated String in case of a Header with
+    * multiple values.
+    */
+  def parse(headerValue: String): Option[A]
+}
+object Header {
+  case class Raw(name: CIString, value: String, recurring: Boolean = true)
+  object Raw {
+    implicit val showForRawHeaders: Show[Raw] =
```

Comment:
We've never gotten consistent about our implicit names.  Some are globally unique in the cats style, some are probably unique like this, and some are just `instance`.

Reply:
Yeah I tried to infer a pattern but failed

Reply:
`catsShowForOrgHttp4sHeadersRaw` would follow the Cats convention, but, geez.

Anyway, that should be a consistency pass before 0.22.  It shouldn't block this.  It's already a mess.

Reply:
I don't really mind the long names, since (hopefully) you are rarely if ever using them explicitly.

Reply:
I'm going to resolve this for now, making implicit names uniform across http4s is probably best left for a different PR :)